solution/server/src/user-data-access.ts

```diff
@@ -0,0 +1,309 @@
+		if (userList.length > limit) {
+			while (userList.length > limit) {
+				userList.pop()
+			}
+
+			// Set the "lastEvaluatedKey" param as the last User retrieved ID to start the next request from it.
+			// We must do that to get the manually removed items
+			const lastIndex = userList.length - 1
+			lastEvaluatedId = userList[lastIndex].id
+		}
+
+		return {
+			users: userList,
+			lastEvaluatedKey: lastEvaluatedId,
+		}
+	} catch (error) {
+		const errorMessage = 'Error on User list'
+		console.log(errorMessage)
+		throw new ApolloError(errorMessage, error)
+	}
+}
+
+/**
+ * Creates an User
+ * @param db The DynamoDB DocumentClient param
+ * @param fields User data
+ * @returns The created user
+ */
+export const AddUserAsync = async (db: DocumentClient, fields: UserInput): Promise<User> => {
+	const addUserData = genUserDataMap(fields) as PutItemInputAttributeMap
+	addUserData['id'] = uuidv4() as AttributeValue
+	addUserData['createdAt'] = addUserData['updatedAt']
+	return AddOrUpdateUserAsync(db, addUserData, true)
+}
+
+/**
+ * Updates an User
+ * @param db The DynamoDB DocumentClient param
+ * @param id The User ID
+ * @param fields User data
+ * @returns The updated User
+ */
+export const UpdateUserAsync = async (db: DocumentClient, id: string, fields: UserInput): Promise<User> => {
+	const addUserData = genUserDataMap(fields) as PutItemInputAttributeMap
+	addUserData['id'] = id as AttributeValue
+	return AddOrUpdateUserAsync(db, addUserData)
+}
+
+/**
+ * Add or Update users
+ * @param db The DynamoDB DocumentClient param
+ * @param data User Data
+ * @returns User
+ */
+const AddOrUpdateUserAsync = async (db: DocumentClient, data: PutItemInputAttributeMap, create = false): Promise<User> => {
+	try {
+		const params: PutItemInput = {
+			TableName: tableName,
+			Item: data,
+		}
+
+		if (create) {
+			params.ConditionExpression = 'attribute_not_exists(id)'
+		} else {
+			params.ConditionExpression = 'attribute_exists(id)'
+		}
+
+		await db
+			.put(params)
+			.promise()
+			.catch((err) => {
+				throw err
+			})
+		return await GetUserByIdAsync(db, data['id'] as string)
+	} catch (error) {
+		const errorMessage = 'Error on User add/update'
+		console.log(errorMessage)
+		throw new ApolloError(errorMessage, error)
+	}
+}
+
+/**
+ * Deletes an User
+ * @param db The DynamoDB DocumentClient param
+ * @param id The User ID
+ * @returns If the user was deleted or not
+ */
+export const DeleteUserAsync = async (db: DocumentClient, id: string): Promise<boolean> => {
+	try {
+		const params: DeleteItemInput = {
+			TableName: tableName,
+			Key: {
+				id: id as AttributeValue,
+			},
+		}
+
+		await db
+			.delete(params)
+			.promise()
+			.catch((err) => {
+				throw err
+			})
+
+		return true
+	} catch (error) {
+		const errorMessage = 'Error on User delete'
+		console.log(errorMessage)
+		throw new ApolloError(errorMessage, error)
+	}
+}
+
+//
+//
+//  Utils
+//
+//
```

Comment:
Do you think that keeping these functions in the same place is better than creating a separate utils file?

Reply:
No, I don't. I usualIy like to put utils functions in another file, in this case, since I would not share this code with other functions or classes, I let them in this file and did not export them, I tried to make them "private". But they can totally be in another file, it would improve project structure, decrease "user-data-access.ts" size (it is already too big) and it would make them testable.